File: web/src/main/java/com/messio/clsb/Position.java
package com.messio.clsb;

import com.messio.clsb.entity.PositionConverter;

import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by jpc on 9/29/16.
 */
public class Position extends HashMap<String, BigDecimal> {
    public static final Position ZERO = new Position();
    public static final PositionConverter CONVERTER = new PositionConverter();

    public static Position parse(String s) {
        return CONVERTER.convertToEntityAttribute(s);
    }

    public Position() {
    }

    public Position(String iso, double amount) {
        this.putAmount(iso, amount);
    }

    public Position(String iso1, double amount1, String iso2, double amount2){
        this.putAmount(iso1, amount1).putAmount(iso2, amount2);
    }

    public Position(String iso1, double amount1, String iso2, double amount2, String iso3, double amount3){
        this.putAmount(iso1, amount1).putAmount(iso2, amount2).putAmount(iso3, amount3);
    }

    public Position(Position p){
        this.putAll(p);
    }

    private Position putAmount(String iso, double amount){
        this.put(iso, new BigDecimal(amount));
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Position){
            final Position that = (Position) o;
            return subtract(that).normalize().isZero();
        }
        return false;
    }

    public boolean isLong(){
        for (BigDecimal amount: normalize().values()){
            if (amount.signum() < 0){
                return false;
            }
        }
        return !isZero();
    }

    public boolean isShort(){
        for (BigDecimal amount: normalize().values()){
            if (amount.signum() > 0){
                return false;
            }
        }
        return !isZero();
    }

    public Position applyVolatility(Position that){
        return entrySet().stream().collect(Position::new, (p, e) -> {
            final BigDecimal vm = that.getOrDefault(e.getKey(), BigDecimal.ZERO);
            final BigDecimal am = e.getValue();
            p.put(e.getKey(), am.signum() > 0 ? am.subtract(am.multiply(vm)): am.add(am.multiply(vm)));
        }, Position::add);
    }

    public Position one(){
        return entrySet().stream().collect(Position::new, (p, e) -> p.put(e.getKey(), BigDecimal.ONE), Position::add);
    }

    public Position xlong(){
        return entrySet().stream().filter(e -> e.getValue().signum() > 0).collect(Position::new, (p, e) -> p.put(e.getKey(), e.getValue()), Position::add);
    }

    public Position xshort(){
        return entrySet().stream().filter(e -> e.getValue().signum() < 0).collect(Position::new, (p, e) -> p.put(e.getKey(), e.getValue()), Position::add);
    }

    public Position normalize(){
        return entrySet().stream().filter(e -> e.getValue().signum() != 0).collect(Position::new, (p, e) -> p.put(e.getKey(), e.getValue()), Position::add);
    }

    public boolean isZero(){
        return normalize().isEmpty();
    }

    public Position add(Position that){
        return entrySet().stream().collect(() -> new Position(that), (p, e) -> p.put(e.getKey(), e.getValue().add(p.getOrDefault(e.getKey(),BigDecimal.ZERO))), Position::add);
    }

    public Position subtract(Position that){
        return add(that.negate());
    }

    public Position negate(){
        return entrySet().stream().collect(Position::new, (p, e) -> p.put(e.getKey(), e.getValue().negate()), Position::add);
    }

    public Position filter(String... isos){
        final List<String> ts = Arrays.asList(isos);
        return entrySet().stream().filter(e -> ts.contains(e.getKey())).collect(Position::new, (p, e) -> p.put(e.getKey(), e.getValue()), Position::add);
    }

    @Override
    public String toString() {
        return CONVERTER.convertToDatabaseColumn(this);
    }
}

File: src/main/kotlin/com/messio/clsb/Position.kt
package com.messio.clsb

import java.math.BigDecimal

class Position: HashMap<String, BigDecimal> {
    constructor(): super()

    constructor(map: Map<String, BigDecimal>) : super(map)

    constructor(vararg amounts: Map.Entry<String, BigDecimal>){
        amounts.forEach { put(it.key, it.value) }
        normalize()
    }

    constructor(vararg amounts: Pair<String, BigDecimal>) {
        amounts.forEach { put(it.first, it.second) }
        normalize()
    }


    private fun putLeg(iso: String, amount: BigDecimal): Position {
        this[iso] = amount
        return this
    }

    fun add(that: Position) = entries.fold(that.copy()){ p, e -> p.putLeg(e.key, e.value.add(p.getOrDefault(e.key, BigDecimal.ZERO)))}.normalize()

    fun subtract(that: Position) = add(that.negate())

    fun negate() = entries.fold(Position()){ p, e -> p.putLeg(e.key, e.value.negate())}

    fun xlong() = Position(filter { it.value.signum() > 0})

    fun xshort() = Position(filter { it.value.signum() < 0})

    fun isZero() = normalize().isEmpty()

    fun isLong(): Boolean = normalize().values.all { it.signum() > 0}

    fun isShort(): Boolean = normalize().values.all { it.signum() < 0 }

    fun normalize() = entries.filter { it.value.signum() != 0 }.fold(Position()){ p, e -> p.putLeg(e.key, e.value) }

    fun copy() = entries.fold(Position()){ p, e -> p.putLeg(e.key, e.value)}

    override fun equals(other: Any?): Boolean = (other as Position).subtract(this).isZero()

    companion object {
        val ZERO = Position()

        fun parse(s: String): Position {
            val position = Position()
            val trim = s.trim()
            if (trim.startsWith("{") && trim.endsWith("}")){
                trim.substring(1, trim.length - 1).split(",")
                    .filter { it.contains("=") }
                    .map { Pair(it.substring(0, it.indexOf("=")).trim().uppercase(), it.substring(it.indexOf("=") + 1))}
                    .filter { it.first.length == 3 &&  it.second.toFloatOrNull() != null }
                    .forEach { position[it.first] = BigDecimal(it.second) }
            }
            return position
        }
    }
}